tfx/benchmarks/beam_pipeline_benchmark_base.py

```diff
@@ -0,0 +1,253 @@
+# Copyright 2019 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Base class for running Apache Beam pipeline based benchmarks."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tfx.benchmarks import mode_config
+from tfx.benchmarks.datasets.chicago_taxi import dataset
+from tfx.benchmarks.tfma_benchmark_chicago_taxi import TFMABenchmarkChicagoTaxi
+from tfx.benchmarks.tfma_v2_benchmark_chicago_taxi import TFMAV2BenchmarkChicagoTaxi
+from tfx.benchmarks.tft_benchmark_chicago_taxi import TFTBenchmarkChicagoTaxi
+from tfx.benchmarks.big_shuffle_benchmark import BigShuffleBenchmarkBase
+
+import subprocess
+import tempfile
+import time
+import yaml
+import csv
+import os
+
+TFMA = "tfma"
+TFMA_V2 = "tfma_v2"
+TFT = "tft"
+BIG_SHUFFLE = "big_shuffle"
+
+TFMA_BENCHMARK_MINI_PIPELINE = "TFMABenchmarkChicagoTaxi.benchmarkMiniPipeline"
+TFMA_V2_BENCHMARK_MINI_PIPELINE_UNBATCHED = "TFMAV2BenchmarkChicagoTaxi.benchmarkMiniPipelineUnbatched"
+TFMA_V2_BENCHMARK_MINI_PIPELINE_BATCHED = "TFMAV2BenchmarkChicagoTaxi.benchmarkMiniPipelineBatched"
+TFT_BENCHMARK_ANALYZE_AND_TRANSFORM_DATASET = "TFTBenchmarkChicagoTaxi.benchmarkAnalyzeAndTransformDataset"
+BIG_SHUFFLE_BENCHMARK = "BigShuffleBenchmarkBase.benchmarkBigShuffle with file size: "
+BIG_SHUFFLE_EMPTY_PIPELINE_BENCHMARK = "BigShuffleBenchmarkBase.benchmarkEmptyPipeline"
+
+
+class BeamPipelineBenchmarkBase():
+  """Beam Pipeline benchmarks base class."""
+
+  def __init__(self, min_num_workers, max_num_workers, base_dir,
```

Comment:
I feel like many of the variables here could be parameters instead of instance variables:

- The runner-specific arguments could be moved to the runner-specific methods e.g. `cloud_dataflow_temp_loc` could be a parameter in `benchmarkCloudDataflow`; that way a user who wants to only run Flink benchmarks would not have to also configure Cloud Dataflow.
- It would be better for `num_workers` to be a local variable that is passed into only the methods that need them.

What do you think? This would require quite a bit of restructuring so I'm OK with leaving it out of this PR.

Reply:
I went through the variables that we discussed yesterday and made as many of them as possible local variables. For `cloud_dataflow_project` and `cloud_dataflow_temp_loc`, I set them as parameters to be passed into `benchmarkCloudDataflow` and removed them from the `BeamPipelineBenchmarkBase` constructor. However, I kept those two variables as instance variables  (i.e. with `self.`) because plumbing those as parameters through the methods that would require them made the code confusing and difficult to understand in my opinion. Let me know if you would like me to make those local as well.

Reply:
This looks OK (since all of this classes' public methods require `big_shuffle_input_file` and `big_shuffle_output_file`.